packages/@eventual/core/src/entity/entity.ts

```diff
@@ -0,0 +1,446 @@
+export type EntityZodShape<Attr extends Attributes> = {
+  [key in keyof Attr]: z.ZodType<Attr[key]>;
+};
+
+/**
+ * An eventual entity.
+ *
+ * @see entity
+ */
+export interface Entity<
+  Attr extends Attributes = any,
+  Partition extends CompositeKeyPart<Attr> = CompositeKeyPart<Attr>,
+  Sort extends CompositeKeyPart<Attr> | undefined =
+    | CompositeKeyPart<Attr>
+    | undefined
+> extends Omit<EntitySpec, "attributes" | "streams" | "partition" | "sort"> {
+  kind: "Entity";
+  key: KeyDefinition;
+  attributes: z.ZodObject<EntityZodShape<Attr>>;
+  streams: EntityStream<Attr, Partition, Sort>[];
+  /**
+   * Get a value.
+   * If your values use composite keys, the namespace must be provided.
+   *
+   * @param key - key or {@link CompositeKey} of the value to retrieve.
+   */
+  get(key: CompositeKey<Attr, Partition, Sort>): Promise<Attr | undefined>;
+  /**
+   * Get a value and metadata like version.
+   * If your values use composite keys, the namespace must be provided.
+   *
+   * @param key - key or {@link CompositeKey} of the value to retrieve.
+   */
+  getWithMetadata(
+    key: CompositeKey<Attr, Partition, Sort>
+  ): Promise<EntityWithMetadata<Attr> | undefined>;
+  /**
+   * Sets or updates a value within an entity and optionally a namespace.
+   *
+   * Values with namespaces are considered distinct from value without a namespace or within different namespaces.
+   * Values and keys can only be listed within a single namespace.
+   */
+  set(entity: Attr, options?: EntitySetOptions): Promise<{ version: number }>;
+  /**
+   * Deletes a single entry within an entity and namespace.
+   */
+  delete(
+    key: CompositeKey<Attr, Partition, Sort>,
+    options?: EntityConsistencyOptions
+  ): Promise<void>;
+  /**
+   * List entries that match a prefix within an entity and namespace.
+   *
+   * If namespace is not provided, only values which do not use composite keys will be returned.
+   */
+  query(
+    key: QueryKey<Attr, Partition, Sort>,
+    request?: EntityQueryOptions
+  ): Promise<EntityQueryResult<Attr>>;
+  stream(
+    name: string,
+    options: EntityStreamOptions<Attr, Partition, Sort>,
+    handler: EntityStreamHandler<Attr, Partition, Sort>
+  ): EntityStream<Attr, Partition, Sort>;
+  stream(
+    name: string,
+    handler: EntityStreamHandler<Attr, Partition, Sort>
+  ): EntityStream<Attr, Partition, Sort>;
+}
+
+export interface EntityTransactItem<
+  Attr extends Attributes = any,
+  Partition extends CompositeKeyPart<Attr> = CompositeKeyPart<Attr>,
+  Sort extends CompositeKeyPart<Attr> | undefined =
+    | CompositeKeyPart<Attr>
+    | undefined
+> {
+  entity: Entity<Attr, Partition, Sort> | string;
+  operation:
+    | EntitySetOperation<Attr>
+    | EntityDeleteOperation<Attr, Partition, Sort>
+    | EntityConditionalOperation<Attr, Partition, Sort>;
+}
+
+export interface EntitySetOperation<Attr extends Attributes = any> {
+  operation: "set";
+  value: Attr;
+  options?: EntitySetOptions;
+}
```

Comment:
Can we move these? There's too much in this file and it takes too long to get to the meat of this file (Options, Entity, entity)

Reply:
Woops, meant to move these to the bottom with the rest of the operation input and outputs. Moved.